README.md
Auteur : Julien CHATEAU

_Ce document est à lire attentivement et vous permettra de comprendre le fonctionnement de ce jeu._

# __SteaMyGame__

## Prérequis

Vous devez posséder la version la plus récente de Java, téléchargeable gratuitement [ici](https://www.java.com/fr/download/)

## L'initialisation et le lancement du jeu

Une fois que vous avez téléchargé l'archive *SteaMyGame-master.zip* via le bouton vert **Clone or download**, il faut tout extraire (**Clique droit sur l'archive** --> **Extraire tout...** --> **Extraire**). Dans le dossier créé, double-cliquer sur le script *installer* qui initialisera le jeu.

Vous n'aurez qu'à supprimer le dossier *SteaMyGame-master* qui est maintenant vide et à double-cliquer sur le raccourcis *SteaMyGame.lnk*, fraichement créé sur votre bureau, pour lancer le jeu !


Si le dossier *SteaMyGame-master* n'est pas vide, l'initialisation n'a pas fonctionner mais vous pouvez quand même lancer le jeu avec le raccourcis *SteaMyGame.lnk* (sans icone)

Si le raccourcis n'est pas créer sur votre bureau, Allez dans *Ce PC*, dans l'onglet *Périphériques et lecteurs*, allez dans *OS (C:)*. Dans le dossier *SteaMyGame*, vous trouverez le raccourcis (avec icone) que vous pourrez récupérer.

Il serait préférable de me contacter (`steamygamecontact@gmail.com`) si une erreur se produit pour que je puisse adapter/corriger mon launcher d'initialisation.

## La compilation

Si vous êtes dans la capacité de modifier le code source, la compilation de l'ensemble des classes se fait ainsi :

`$javac -cp bin -d bin -encoding UTF-8 src/X/*.java`  
(avec X le répertoire dans lequel se trouve la(les) classe(s) que vous avez modifiée(s))

Pour compiler tous les fichiers d'un coup (possible seulement sur Linux) :  
1. `$shopt -s globstar`
2. `$javac -cp bin -d bin -encoding UTF-8 src/**/*.java`

## Les mécaniques du jeu

   Pour jouer à ce jeu vous devez vous créer un compte avec un login et un mot de passe (à ne pas oublier).
Lorsque vous vous connectez à votre compte pour la première fois, vous devez choisir un personnage parmi les 32 disponibles.

**/!\\ VOUS NE POUVEZ PAS CHANGER DE PERSONNAGE APRÈS VOTRE CHOIX /!\\**

Après une longue réflexion (aidée par le guide des personnages plus bas) et après avoir fait votre choix, vous arrivez sur l'interface de l'inventaire.
Le jeu peut commencer !!


### Synopsis, déroulement et règles du jeu
Après un voyage de plusieurs jours, vous arrivez devant la gigantesque **Tour Celeste**. Vous avez été envoyé par le roi de votre village pour vaincre Imleryth,
le démon de glace, qui est sur le point de ravager votre monde depuis le haut de cette tour. Cependant cela ne va pas être chose facile car il vous
faudra d'abord gravir les 50 étages de cette tour. De plus, durant votre voyage vous avez été attaqué par un groupe de bandit qui vous a dérobé tout votre
équipement. Vous allez donc commencer votre ascenssion sans rien dans les poches (ou peut-être pas). Chaque étages est composées de 3 salles et d'un monstre par salle.
Plus vous avancerez dans la tour, plus les monstres seront fort. Les combats se déroulent en tour par tour.


### La Tour :
* Dès que vous battez un monstre vous gagnez des **$teamy**, l'argent du jeu.
* Si vous perdez ou abandonnez un combat, vous perdez de l'argent et devez recommencer tout l'étage.
* Si vous battez les 3 monstres d'un étage à la suite, vous gagnez une rune (voir explication sur les runes plus bas).
* Entre chaque salle vous regagnez 75% de vos points de vie manquant.
* Tous les 10 étages les monstres sont plus fort que la normale et vous gagnerez une rune supplémentaire à la première victoire.
* Vous ne pouvez refaire que les 4 derniers étages que vous avez vaincu et gagnerez deux fois moins de $teamy.


### Les Runes :
* Vous pouvez équiper jusqu'a 5 runes.
* Vous pouvez vendre une rune contre des $teamy.
* Une rune peut être positionnée de 5 façons différentes (en position 0, 1, 2, 3 ou 4).
* Il existe 6 types de runes différents ayant leur propre bonus :
    * **Fatale** (+20% d'attaque)
    * **Energie** (+20% de points de vie)
    * **Gardien** (+20% de défense)
    * **Swift** (+22% de vitesse)
    * **Rage** (+26% de dégats critiques)
    * **Lame** (+12% de taux critiques)
* Si une rune est en position 0, le bonus est activé une fois.
* Sinon le bonus est activé en équipant 2 runes du même type sur les autres positions (cumulable 2 fois => 3 fois max en tout).
* Dans l'inventaire une rune aura un affichage du style : TypePosition (ex : R3 décrit une rune Rage de position 3)
* Une rune possède un stat principale (détail de la stat principal en fonction de la position plus bas) et de 1 à 4 stats secondaires.
* Les stats existantes (flat et/ou en %) sont :
    * **L'attaque** (flat et en %) --> Att/Att%
    * **La défense** (flat et en %) --> Def/Def%
    * **Les points de vie** (flat et en %) --> Hp/Hp%
    * **La vitesse** (flat) --> Vit
    * **Les taux de critiques** --> Tx crit.
    * **Les dégats critiques** (en %) --> Dgts crit.
* Initialement un rune possède 1 stat principale et 1 stat secondaire.
* Un rune possède aussi un niveau allant de 1 à 20 et peut être amélioré contre des $teamy.
* Chaque amélioration augmente un peu la stat principale.
* Tout les 5 niveaux vous gagnez une stat secondaire aléatoire supplémentaire.
* Si vous atteignez le niveau maximum d'une rune, totues ses stats secondaires seront augmentées de 50%.
* Les stats principales possibles en fonction de la position de la rune :
    * **0** : Toutes
    * **1** : Att, Hp, Def
    * **2** : Att%, Hp%, Def%, Vit
    * **3** : Tx crit, Dgts crit, Att%, Hp%, Def%
    * **4** : Att, Hp, Def, Att%, Hp%, Def%


### Les Personnages :
* Un personnage possède les 6 caractéristiques précédement vu.
* La valeur de ses caractéristiques varie en fonction du type du personnage.
* Un personnage possède 4 sorts ayant un cooldown (nombre de tours sans pouvoir utiliser un sort après utilisation) qui se réduit chaque tour.
* Le sort n°1 n'a pas de cooldown et le sort n°4 peut être un passif (non lançable) pour certain personnage.
* Les sorts peuvent appliquer à l'ennemi des effets nocifs et vous procurer des effets bénéfiques, détaillés ci dessous.


### Les Effets Nocifs (appliqué pendant 1 à 3 tours sur la cible) :
* **Break Att** : Réduit l'attaque.
* **Break Def** : Réduit la défense.
* **Break Spd** : Réduit la vitesse.
* **Anti-Heal** : Empèche de se procurer toute sorte de soin.
* **Marque** : Augmente les dégats reçu de 25%.
* **Silence** : Oblige l'utilisation du sort n°1.
* **Etourdissement** : Empèche de jouer.
* **Bombe** : 2 tours après application, inflige des dégats et étourdis avec 35%.
* **Degats Continus** : Chaque tour, inflige des dégats équivalents à 5% des points de vie.

### Les Effets Bénéfiques (appliqué pendant 1 à 3 tours sur la cible):
* **Buff Att** : Augmente l'attaque.
* **Buff Def** : Augmente la défense.
* **Buff Spd** : Augmente la vitesse.
* **Buff Crit** : Augmente les taux critiques.
* **Immunité** : Vous résistez à tout effets nocifs.
* **Invincibilité** : Vous ne subissez aucun dégat.
* **Contre-Attaque** : Si vous subissez des dégats (directs) de l'ennemi, vous contre-attaquez avec votre sort n°1.

PS : Si vous remarquez quelque chose semblant être erroné n'hésitez pas à me contacter, je me ferai un plaisir de le corriger
`steamygamecontact@gmail.com`


CHANGELOG.md
Auteur : Julien CHATEAU

# JOURNAL DES MISES À JOUR

## 25/04/2018
* Changement de l'expression d'un sort en cooldown
    * Avant : Lorsque un sort en cooldown était choisi, un message était afficher dans le chat
    * Mantenant : Les sorts en cooldown sont désactivé et leur couleur devient gris/rouge

## 02/05/2018
* Correction d'un bug à cause duquel les boutons des sorts passifs étaient activables
* Correction d'un bug à cause duquel la couleur du bouton du sort n°4 restait gris/rouge alors que le cooldown était à 0

## 08/05/2018
* Les caractéristiques de tous les personnages de la tour ont été diminuée de 10%
* L'argent gagné lors d'une victoire d'un étage a été augmenté
* Nephthys : Les dégats du skill n°2 on été légerement baissés
* Thrain : correction d'un bug à cause duquel le nombre de tour du boost d'attaque accordé par le skill n°4 de correspondait pas au nombre d'ennemis étourdis
* Tablo : correction d'un bug à cause duquel les effets bénéfiques procurés par le passif était appliqué sur les ennemis et non sur les alliés

## 13/05/2018
* Eshir :
    * Le passif n'accorde plus que des dégats supplémentaire en fonction de la différence de points de vie entre eshir et la cible
    * Les dégats du sort n°3 dépendent un peu moins des points de vie maximum mais plus de l'attaque.

## 04/09/2018
* Correction de certaines fautes d'orthographes et de certaines génération de texte incorrecte dans les parties narratives du jeu

## 05/09/2018
* Thebae :
    * Le sort n°3 soigne correctement de 50% des dégats infligés et non de 75%
    * Chacune des attaques du sort n°2 n'a plus que 35% de chance d'appliquer un marque sur l'ennemi (50% auparavant)
* Artamiel :
    * La description du sort n°2 a été ajusté
    * Le montant de soin procuré par le sort n°3 passe de 55% à 50% des points de vie maximums de l'allié ciblé.
    * Le sort n°4 augmente désormait la vitesse des alliés pendant 1 tour en plus des précédentes fonctionnalités
    * Une partie des dégats du sort n°1 dépende de la défense d'Artamiel
* Chasun :
    * L'IA du sort n°4 a été amélioré

## 06/09/2018
* Correction d'un bug à cause duquel lorsque l'on choisissait un sort autre que le n°1 en étant sous effet de silence, l'alerte "*\*Nom_Du_Personnage\** est silence !!" dans le chat n'apparraissait pas.
* Correction d'un bug à cause duquel un personnage KO pouvait recevoir des points de bouclier
* Thebae :
    * L'affichage de la resistance de l'effet de Marque du sort n°2 a été factoriser pour une meilleur visibilité dans le chat

## 08/09/2018
* Nephtys :
    * Correction d'un bug à cause duquel les effets du sort n°3 passait à travers l'immunité
* Seara :
    * Correction d'un bug à cause duquel Seara était étourdis après le lancement de sont sort n°4 alors qu'elle possedait une immunité

## 10/09/2018
* Nephtys :
    * Correction d'un bug d'affichage dans le chat lors du lancement du sort n°3


src/autre/TourCeleste.java
package comptes;
import personnage.*;
import java.util.*;
import java.io.*;
import java.nio.file.*;
import application.*;


public class TourCeleste {

  private Etage[] tour;
  private GestionFichiers init;

  public TourCeleste() {
    this.tour = new Etage[35];
    try {
      this.init = new GestionFichiers(new File("data/RunesTourCeleste.txt"));
      Personnage kaito1 = new Kaito("");
      Personnage ritesh1 = new Ritesh("");
      Personnage zaiross1 = new Zaiross("");

      Personnage basalt2 = new Basalt("");
      Personnage tablo2 = new Tablo("");
      Personnage eshir2 = new Eshir("");

      Personnage sige10 = new Sige("");
      Personnage haken10 = new Haken("");
      Personnage cichlid10 = new Cichlid("");
      Personnage cadiz10 = new Cadiz("");
      Personnage erwin10 = new Erwin("");
      Personnage nicki10 = new Nicki("");

      Personnage eshir20 = new Eshir("");
      Personnage nicki20 = new Nicki("");
      Personnage icaru20 = new Icaru("");

      Personnage tablo30 = new Tablo("");
      Personnage eludia30 = new Eludia("");
      Personnage woonhak30 = new Woonhak("");

      Personnage veromos40 = new Veromos("");
      Personnage zaiross40 = new Zaiross("");
      Personnage stella40 = new Stella("");
      Personnage cichlid40 = new Cichlid("");
      Personnage kaito40 = new Kaito("");
      Personnage nephthys40 = new Nephthys("");

      Personnage cadiz50 = new Cadiz("");
      Personnage cuivre50 = new Cuivre("");
      Personnage seara50 = new Seara("");

/*
* REMPLISSAGE DE LA TOUR
*/


      this.tour[0] = new Etage(kaito1.malusPremiersNiveaux(12), ritesh1.malusPremiersNiveaux(12), zaiross1.malusPremiersNiveaux(12));
      this.tour[1] = new Etage(basalt2.malusPremiersNiveaux(6), tablo2.malusPremiersNiveaux(6), eshir2.malusPremiersNiveaux(6));
      this.tour[2] = new Etage(new Fuco(""), new Veromos(""), new Okeanos(""), new Sian(""), new Zeratu(""), new Seara(""));//1 rune --- 2 runes
      this.tour[3] = new Etage(new Cuivre(""), new Eludia(""), new Thrain(""), new Woonhak(""), new XiongFei(""), new Icaru(""));//3 runes --- 3 runes
      this.tour[4] = new Etage(new XiaoLin(""), new Oberon(""), new Nephthys(""));//4 runes
      this.tour[5] = new Etage(new Stella(""), new Ethna(""), new Thebae(""));//4 runes
      this.tour[6] = new Etage(cadiz10.boostEtageBoss(10), erwin10.boostEtageBoss(10), nicki10.boostEtageBoss(10), sige10.boostEtageBoss(10), cichlid10.boostEtageBoss(10), haken10.boostEtageBoss(10));//4 runes --- 5 runes


      this.tour[7] = new Etage(new Fuco(""), new Cuivre(""), new XiaoLin(""), new Stella(""), new Kaito(""), new Chasun(""));
      this.tour[8] = new Etage(new Haken(""), new Artamiel(""), new Thrain(""));
      this.tour[9] = new Etage(new Eludia(""), new Oberon(""), new Ethna(""), new Nephthys(""), new Zaiross(""), new Cichlid(""));
      this.tour[10] = new Etage(new Sian(""), new Woonhak(""), new Chasun(""));
      this.tour[11] = new Etage(new Cadiz(""), new Zeratu(""), new XiongFei(""), new Tablo(""), new Erwin(""), new Seara(""));
      this.tour[12] = new Etage(new Thebae(""), new Ritesh(""), new Sige(""), new Chasun(""), new Basalt(""), new Veromos(""));
      this.tour[13] = new Etage(icaru20.boostEtageBoss(20), eshir20.boostEtageBoss(20), nicki20.boostEtageBoss(20));


      this.tour[14] = new Etage(new Ethna(""), new Erwin(""), new Basalt(""), new Fuco(""), new Cadiz(""), new Zaiross(""));
      this.tour[15] = new Etage(new Zeratu(""), new XiongFei(""), new Sige(""), new Icaru(""), new Nephthys(""), new Stella(""));
      this.tour[16] = new Etage(new Oberon(""), new Veromos(""), new Okeanos(""));
      this.tour[17] = new Etage(new Kaito(""), new Nicki(""), new XiaoLin(""));
      this.tour[18] = new Etage(new Seara(""), new Haken(""), new Okeanos(""), new Artamiel(""), new Cuivre(""), new Sian(""));
      this.tour[19] = new Etage(new Cichlid(""), new Thrain(""), new Ritesh(""));
      this.tour[20] = new Etage(tablo30.boostEtageBoss(30), eludia30.boostEtageBoss(30), woonhak30.boostEtageBoss(30));


      this.tour[21] = new Etage(new Eludia(""), new Woonhak(""), new Seara(""), new Sian(""), new Erwin(""), new Ethna(""));
      this.tour[22] = new Etage(new Tablo(""), new Oberon(""), new Basalt(""), new Icaru(""), new Ritesh(""), new XiongFei(""));
      this.tour[23] = new Etage(new Thebae(""), new Zeratu(""), new Nicki(""), new Cadiz(""), new Eshir(""), new Sige(""));
      this.tour[24] = new Etage(new Thrain(""), new Cuivre(""), new Haken(""));
      this.tour[25] = new Etage(new XiaoLin(""), new Okeanos(""), new Fuco(""));
      this.tour[26] = new Etage(cichlid40.boostEtageBoss(40), kaito40.boostEtageBoss(40), nephthys40.boostEtageBoss(40), veromos40.boostEtageBoss(40), zaiross40.boostEtageBoss(40), stella40.boostEtageBoss(40));


      this.tour[27] = new Etage(new Thebae(""), new Oberon(""), new Stella(""));
      this.tour[28] = new Etage(new Icaru(""), new Ritesh(""), new Nephthys(""), new XiaoLin(""), new Woonhak(""), new Artamiel(""));
      this.tour[29] = new Etage(new Sian(""), new Erwin(""), new Haken(""));
      this.tour[30] = new Etage(new Eludia(""), new Okeanos(""), new Sige(""), new Veromos(""), new Kaito(""), new Nicki(""));
      this.tour[31] = new Etage(new Cichlid(""), new Zaiross(""), new XiongFei(""), new Tablo(""), new Zeratu(""), new Basalt(""));
      this.tour[32] = new Etage(new Thrain(""), new Eshir(""), new Ethna(""));
      this.tour[33] = new Etage(cadiz50.boostEtageBoss(50), cuivre50.boostEtageBoss(50), seara50.boostEtageBoss(50));


      this.tour[34] = new Etage(new Chasun(""), new Artamiel(""), new Woonhak(""), new Cichlid(""), new Thebae(""), new Imleryth(""));


/*
* EQUIPEMENT DES RUNES DE TOUS LES PERSOS DE LA TOUR EN LISANT LE FICHIER DE SAUVEGARDE
*/
      int etagesdispo = 35;
      this.init.ouvrir();
      for (int i = 2; i<etagesdispo; i++) {
        for (int j = 1; j<7; j++) {
          if (this.tour[i].getPersoSalle(j) != null) {
            if (i == 2 && j < 4) {
              this.tour[i].getPersoSalle(j).equiperRune(Application.stringToRune(this.init.lire()));
            } else
            if (i == 2 && j >= 4) {
              this.tour[i].getPersoSalle(j).equiperRune(Application.stringToRune(this.init.lire()));
              this.tour[i].getPersoSalle(j).equiperRune(Application.stringToRune(this.init.lire()));
            } else
            if (i == 3) {
              this.tour[i].getPersoSalle(j).equiperRune(Application.stringToRune(this.init.lire()));
              this.tour[i].getPersoSalle(j).equiperRune(Application.stringToRune(this.init.lire()));
              this.tour[i].getPersoSalle(j).equiperRune(Application.stringToRune(this.init.lire()));
            } else
            if (i == 4 || i == 5 || (i == 6 && j < 4)) {
              this.tour[i].getPersoSalle(j).equiperRune(Application.stringToRune(this.init.lire()));
              this.tour[i].getPersoSalle(j).equiperRune(Application.stringToRune(this.init.lire()));
              this.tour[i].getPersoSalle(j).equiperRune(Application.stringToRune(this.init.lire()));
              this.tour[i].getPersoSalle(j).equiperRune(Application.stringToRune(this.init.lire()));
            } else {
              this.tour[i].getPersoSalle(j).equiperRune(Application.stringToRune(this.init.lire()));
              this.tour[i].getPersoSalle(j).equiperRune(Application.stringToRune(this.init.lire()));
              this.tour[i].getPersoSalle(j).equiperRune(Application.stringToRune(this.init.lire()));
              this.tour[i].getPersoSalle(j).equiperRune(Application.stringToRune(this.init.lire()));
              this.tour[i].getPersoSalle(j).equiperRune(Application.stringToRune(this.init.lire()));
            }
          }
        }
      }

      this.init.fermer();

      // // AFFICHE LES RUNES DE TOUS LES PERSOS DE LA TOUR
      // for (int i = 0; i<etagesdispo; i++) {
      //   for (int j = 1; j<7; j++) {
      //     if (this.tour[i].getPersoSalle(j) != null) {
      //       System.out.println(this.tour[i].getPersoSalle(j).toStringRunes());
      //       System.out.println("----------------");
      //     }
      //   }
      //   System.out.println("___________________________________");
      // }

      Personnage p;
      for (int i = 0; i<etagesdispo; i++) {
       for (int j = 1; j<7; j++) {
         p = this.tour[i].getPersoSalle(j);
         if (p != null) {
           p.malusParceQueTropFort();
         }
       }
     }





/*
* APPLICATION DES EFFETS DES RUNES DE TOUS LES PERSOS
*/
      for (int i = 2; i<etagesdispo; i++) {
        for (int j = 1; j<7; j++) {
          if (this.tour[i].getPersoSalle(j) != null) {
            this.tour[i].getPersoSalle(j).appliquerEffetsRunes();
            this.tour[i].getPersoSalle(j).reajusteCarac();
          }
        }
      }
    } catch (IOException e) {
      System.out.println("Erreur Tour : "+e.getMessage());
    }
  }//fin constructeur

  public Personnage getPerso(int etage, int salle) {
    return this.tour[etage-1].getPersoSalle(salle);
  }

  public Etage getEtage(int etage) {
    return this.tour[etage-1];
  }

  public void resetEtage(int etage) {//permet de reset les caractéristiques des persos de l'étage
    this.tour[etage-1].reset();
  }
}
